### PR TITLE
Started on refract conversion

### PR DESCRIPTION
**Pull request by pksunkara**
This PR starts the refract conversion and expands more on the refract option. It also streamlines a bit of visitor application code.

### CODE REVIEW

File: src/RefractAPI.cc [hunk 1]
```diff
@@ -13,11 +13,154 @@
...
+
+        ast->element("category");
+        ast->meta["classes"] = CreateArrayElement("api");
+        ast->meta["title"] = refract::IElement::Create(blueprint.name);
+
+        if (!blueprint.description.empty()) {
+            elements.push_back(CopyToRefract(blueprint.description));
+        }
+
+        // FIXME: Correct this according to spec
```
Comment (klokane) [inline src/RefractAPI.cc]: remove fixme

Comment (klokane) [inline src/RefractAPI.cc]: will do :)

File: src/RefractAPI.cc [hunk 2]
```diff
@@ -13,11 +13,154 @@
...
+        if (!blueprint.metadata.empty()) {
+            ast->attributes["meta"] = MetadataCollectionToRefract(blueprint.metadata);
+        }
+
+        // Append set of elements to content
+        const snowcrash::Elements& scElements = blueprint.content.elements();
+        std::transform(scElements.begin(), scElements.end(), std::back_inserter(elements), ElementToRefract);
+
+        // Remove NULL elements
```
Comment (klokane) [inline src/RefractAPI.cc]: Refactoring - (repeated on line no. 115) -> extract to function

File: src/RefractAPI.cc [hunk 3]
```diff
@@ -13,11 +13,154 @@
...
+            return ptr == NULL;
+        }
+
+    }
+
+    refract::IElement* MetadataToRefract(const snowcrash::Metadata metadata)
+    {
+        refract::MemberElement* element = new refract::MemberElement;
+
+        element->meta["classes"] = CreateArrayElement("user");
```
Comment (klokane) [inline src/RefractAPI.cc]: Move all const text into `namespace key` like in ` RefractDataStructure.cc`

**Comment by pksunkara** [comment]
Pushed new commit.

File: src/RefractDataStructure.cc [hunk 4]
```diff
@@ -719,6 +719,10 @@ namespace drafter
 
     refract::IElement* DataStructureToRefract(const snowcrash::DataStructure& dataStructure)
     {
+        if (dataStructure.empty()) {
+            return NULL;
+        }
+
```
Comment (pksunkara) [inline src/RefractDataStructure.cc]: @klokane The issue which we need to fix here is that we do not have an [API namespace DataStructure](https://github.com/refractproject/refract-spec/blob/968cf6042757a365115ce88a5ddab6912e0e4a1e/namespaces/api-description-namespace.md#data-structure-element) here. Our AST directly embeds the MSON Refract.

File: src/RefractAPI.cc (hunk 3 above)
Comment (pksunkara) [inline src/RefractAPI.cc]: I will do this in a separate PR.

**Comment by pksunkara** [comment]
Updated the PR. Please review & merge this first.

File: src/RefractAPI.cc (hunk 3 above)
Comment (klokane) [inline src/RefractAPI.cc]: ok

File: src/RefractAPI.cc [hunk 5]
```diff
@@ -7,17 +7,442 @@
...
+        }
+        else {
+            element->element("httpRequest");
+            element->attributes["method"] = refract::IElement::Create(method);
+
+            if (payload != NULL) {
+                element->attributes["title"] = refract::IElement::Create(payload->name);
+            }
+
+            // FIXME: Expand href
```
Comment (klokane) [inline src/RefractAPI.cc]: What does mean this?
It is for this release?

Comment (pksunkara) [inline src/RefractAPI.cc]: I don't think API Blueprint uses this. But I left a remark here so that I can ask the guys later.

**Comment by klokane** [comment]
@pksunkara pls check memory, `valgrind` report some memory leaks, all comes from  `ParameterToRefract`.

Probably some unassigned allocation

File: src/RefractAPI.cc (hunk 5 above)
Comment (klokane) [inline src/RefractAPI.cc]: Ok so please add some little bit descriptive info.

**Comment by pksunkara** [comment]
Updated.

File: src/RefractAPI.cc [hunk 6]
```diff
@@ -7,17 +7,450 @@
...
+        }
+        else {
+            value = refract::IElement::Create(parameter.exampleValue);
+            defaultValue = refract::IElement::Create(parameter.defaultValue);
+        }
+
+        // Values
+        if (!parameter.values.empty()) {
+            if (parameter.type == "boolean") {
+                value = ParameterValuesToRefract<bool>(parameter);
```
Comment (klokane) [inline src/RefractAPI.cc]: probably memory leak - there will be content of `value` lost if it was already set on line  108. dtto for all other assigns below.

File: src/RefractAPI.cc [hunk 7]
```diff
@@ -7,17 +7,450 @@
...
+
+        element->set(content);
+
+        return element;
+    }
+
+    refract::IElement* ParameterToRefract(const snowcrash::Parameter& parameter)
+    {
+        refract::MemberElement* element = new refract::MemberElement;
+        refract::IElement *value, *defaultValue;
```
Comment (klokane) [inline src/RefractAPI.cc]: if declare pointee variables to be assigned later every time init them to NULL

File: src/RefractAPI.cc [hunk 8]
```diff
@@ -7,17 +7,450 @@
...
+                value = ParameterValuesToRefract<std::string>(parameter);
+            }
+        }
+
+        // Default Value
+        if (!parameter.defaultValue.empty()) {
+            value->attributes["default"] = defaultValue;
+        }
+        else {
+            delete defaultValue;
```
Comment (klokane) [inline src/RefractAPI.cc]: it is better to do not allocate memory if it is required to delete it later in same function

**Comment by pksunkara** [comment]
Fixed.

File: src/RefractAPI.cc [hunk 9]
```diff
@@ -7,17 +7,452 @@
...
+            defaultValue = refract::IElement::Create(LiteralTo<double>(parameter.defaultValue));
+        }
+        else {
+            value = refract::IElement::Create(parameter.exampleValue);
+            defaultValue = refract::IElement::Create(parameter.defaultValue);
+        }
+
+        // Values
+        if (!parameter.values.empty()) {
+            delete value;
```
Comment (klokane) [inline src/RefractAPI.cc]: What if value is not set? You will delete NULL

Comment (pksunkara) [inline src/RefractAPI.cc]: That won't happen because of the if-else loop above.

Comment (klokane) [inline src/RefractAPI.cc]: In both cases you are checking `parameter.type` against literal value
What about to join them together?

Something like:
```
     if (parameter.type == "boolean") { 
            if (!parameter.values.empty()) { 
                value = ParameterValuesToRefract<bool>(parameter);
            } 
            else { 
               value = refract::IElement::Create(LiteralTo<bool>(parameter.exampleValue));
            } 
            if (!parameter.defaultValue.empty()) { 
                value->attributes["default"] = refract::IElement::Create(LiteralTo<bool>(parameter.defaultValue));
            } 

        } 
  else if (...) {...}
```

and to DRY you can move this part to template function

**Comment by pksunkara** [comment]
Updated.

**Comment by pksunkara** [comment]
Updated.

**Comment by klokane** [comment]
:+1:

**Pull request merged**